git/rev_list_scanner.go

```diff
@@ -0,0 +1,268 @@
+package git
+
+import (
+	"bufio"
+	"encoding/hex"
+	"io"
+	"io/ioutil"
+	"os/exec"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/git-lfs/git-lfs/errors"
+	"github.com/rubyist/tracerx"
+)
+
+// ScanningMode is a constant type that allows for variation in the range of
+// commits to scan when given to the `*git.RevListScanner` type.
+type ScanningMode int
+
+const (
+	// ScanRefsMode will scan between two refspecs.
+	ScanRefsMode ScanningMode = iota
+	// ScanAllMode will scan all history.
+	ScanAllMode
+	// ScanLeftToRemoteMode will scan the difference between "left" and a
+	// remote tracking ref.
+	ScanLeftToRemoteMode
+)
+
+// ScanRefsOptions is an "options" type that is used to configure a scan
+// operation on the `*git.RevListScanner` instance when given to the function
+// `NewRevListScanner()`.
+type ScanRefsOptions struct {
+	// Mode is the scan mode to apply, see above.
+	Mode ScanningMode
+	// Remote is the current remote to scan against, if using
+	// ScanLeftToRemoveMode.
+	Remote string
+	// SkipDeletedBlobs specifies whether or not to traverse into commit
+	// ancestry (revealing potentially deleted (unreferenced) blobs, trees,
+	// or commits.
+	SkipDeletedBlobs bool
+
+	// SkippedRefs provides a list of refs to ignore.
+	SkippedRefs []string
+	// Mutex guards names.
+	Mutex *sync.Mutex
+	// Names maps Git object IDs (encoded as hex using
+	// hex.EncodeString()) to their names, i.e., a directory name
+	// (fully-qualified) for trees, or a pathspec for blob tree entries.
+	Names map[string]string
+}
+
+// GetName returns the name associated with a given blob/tree sha and "true" if
+// it exists, or ("", false) if it doesn't.
+//
+// GetName is guarded by a use of o.Mutex, and is goroutine safe.
+func (o *ScanRefsOptions) GetName(sha string) (string, bool) {
+	o.Mutex.Lock()
+	defer o.Mutex.Unlock()
+
+	name, ok := o.Names[sha]
+	return name, ok
+}
+
+// SetName sets the name associated with a given blob/tree sha.
+//
+// SetName is guarded by a use of o.Mutex, and is therefore goroutine safe.
+func (o *ScanRefsOptions) SetName(sha, name string) {
+	o.Mutex.Lock()
+	defer o.Mutex.Unlock()
+
+	o.Names[sha] = name
+}
+
+// RevListObject is a tuple type returning results from the RevListScanner.
+type RevListObject struct {
+	// Name is an optional field that gives the name of the object (if the
+	// object is a tree, blob).
+	Name string
+	// Oid is the hex-decoded bytes of the object's ID.
+	Oid []byte
+}
```

Comment:
I don't know if this type is necessary. What do you think about just adding `Name()` and `OID()` funcs on the `RevListScanner`?